### PR TITLE
Apply logic to wait for new workers to be loaded

### PR DESCRIPTION
**Pull request by kevindew**
This tries to resolve [an 8 year old TODO](https://github.com/gds-operations/unicornherder/blob/bd3b5688896731947a0f8fd85b2ad46ba7899e2c/unicornherder/herder.py#L264) by applying some logic beyond just sleeping for when this tool is waiting for new workers to be initialised.

Previously, when switching between unicorn master processes there would be a 2 minute wait for the new worker processes to be created. This replaces that wait with a loop that checks whether the new process has the same number of workers as the old process. If it does it stops waiting and starts shutting down the old process.

It repurposes the overlap argument as a means to specify how long both versions of the app should run together once the new workers have been forked. In unicorn instances when the app is not pre-loaded there can be a delay while the application boots where you'd like the old app to still be serving requests. It is my understanding that during this
period of time either version of the app can be attempting to serve requests so this is kept relatively low at 15 seconds default, with the expectation that if you have an app that needs more than 15 seconds then you should be using the preload_app functionality.

Should the same number of workers not return on the new process the same 2 minute will expire and the new master process would be shut down as before. This could happen in valid circumstances, such as someone reducing the number of workers in a deploy, or if someone had manually upped the old process workers with TTOU signals.

This has been done to reduce the amount of time that two applications are running simultaneously, this prolonged period can cause problems as any monitoring to check whether a deploy is ok has to wait for this to change - or risks checking with a 50/50 chance of which version of an app to be served.

### CODE REVIEW

**Comment by nickstenning** [comment]
Can I just pop in to say how happy it makes me to see someone tackling my lazy 8-year-old TODO...? 😄

**Comment by kevindew** [comment]
> Can I just pop in to say how happy it makes me to see someone tackling my lazy 8-year-old TODO...? 😄

Haha, nice of you to pop by - let's hope we do it justice with so many years of anticipation. This must be a rather surprising blast from the past for you.

File: unicornherder/herder.py [hunk 1]
```diff
@@ -260,10 +260,28 @@ def _emergency_slaughter():
...
-def _wait_for_workers(overlap):
-    # TODO: do something smarter here
-    time.sleep(overlap)
-
+def _wait_for_workers(overlap, new_process, old_process):
+    # We expect the current process has one extra child (the new process that
+    # was forked aside from the usual number of workers
+    current_workers = len(old_process.children()) - 1
+    # We hope for same number of workers, if we don't have that we'll accept 1
+    expected_children = current_workers or 1
```
Comment (thomasleese) [inline unicornherder/herder.py:268]: It seems like this is only relevant if `current_workers` is set to `0`? What situation do we anticipate that happening, is it if the old process died unexpectedly?

And if so, I wonder if we should do `max(current_workers, 1)` to catch a situation where `len(old_process.children())` is `0` and therefore `current_workers` is `-1`?

File: test/test_herder.py [hunk 2]
```diff
@@ -157,12 +220,13 @@ def test_loop_detects_pidchange(self, open_mock, process_mock, sleep_mock):
...
     @patch('%s.open' % builtin_mod)
-    def test_loop_reload_pidchange_signals(self, open_mock, process_mock,
-                                           sleep_mock):
+    def test_loop_reload_pidchange_signals(self, open_mock, process_mock, sleep_mock):
         proc1 = MagicMock()
         proc2 = MagicMock()
         proc1.pid = 123
         proc2.pid = 456
+        proc1.children = MagicMock(return_value=[proc2, "worker"])
+        proc2.children = MagicMock(return_value=["worker 1"])
```
Comment (thomasleese) [inline test/test_herder.py:229]:
```suggestion
        proc1.children = Mock(return_value=[proc2, "worker"])
        proc2.children = Mock(return_value=["worker 1"])
```

I don't think these need to be `MagicMock`s.

Comment (thomasleese) [reply test/test_herder.py:229]: Hmm actually there's a lot of uses of `MagicMock` all across this code where it's not required... so it doesn't matter too much

**Review by thomasleese — COMMENTED** [review]
Great to see this being resolved! I've just got one question, but apart from that it looks good!

File: unicornherder/herder.py (hunk 1 above)
Comment (kevindew) [reply unicornherder/herder.py:268]: That sounds a good idea, it was mostly just defensive as waiting for anything less than one didn't seem logical.

File: test/test_herder.py (hunk 2 above)
Comment (kevindew) [reply test/test_herder.py:229]: Sure, I had no idea that MagicMock and Mock were different things 🤷

**Comment by kevindew** [comment]
@thomasleese I've made that change now. Do you think I should bump the version number in this PR too to save a second one?

I tried this out on integration yesterday (which ended up with me modifying it to add the overlap in because most of our workers are not preloaded) and seemed to work well.

File: test/test_herder.py (hunk 2 above)
Comment (thomasleese) [reply test/test_herder.py:229]: `MagicMock` is only relevant if you're needing to use so called "magic methods" which is broadly what other languages would call operator overloading.

**Comment by thomasleese** [comment]
@kevindew Yeah, happy for the version bump to be included in this PR.

**Comment by kevindew** [comment]
@thomasleese thanks done, though CI hasn't run here you can see it on my branch: https://github.com/kevindew/unicornherder/runs/1040860556

Do you know how to release this after / would you mind doing it for me 🙏

**Review by thomasleese — APPROVED** [review]

**Pull request merged**